wasm/src/logging.rs

```diff
@@ -0,0 +1,168 @@
+use core::sync::atomic::{AtomicU8, Ordering};
+
+use super::*;
+
+// NOTE: `u8::MAX` is a sentinel value for an undefined log level
+static MAX_LOG_LEVEL: MaxLogLevel = MaxLogLevel(AtomicU8::new(u8::MAX));
+
+/// Struct which holds a valid [`LogLevel`] stored as an integer
+struct MaxLogLevel(AtomicU8);
+
+/// Log level supported by the host
+// NOTE: This struct must be exact duplicate of `config::logger::Level`
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord)]
+#[allow(clippy::upper_case_acronyms)]
+#[repr(u8)]
+pub enum LogLevel {
+    /// Trace
+    TRACE,
+    /// Debug
+    DEBUG,
+    /// Info (Default)
+    #[default]
+    INFO,
+    /// Warn
+    WARN,
+    /// Error
+    ERROR,
+}
+
+impl MaxLogLevel {
+    fn get(&self) -> LogLevel {
+        let mut log_level = self.0.load(Ordering::SeqCst);
+
+        if log_level == u8::MAX {
```

Comment:
an alternative to lazy loading would be to query host log level at the beginning of every smart contract

Reply:
So right now we query this at the first attempting to log?

Reply:
yes

Reply:
imo such lazy evaluation is fine, why load anything if it won't be used

Reply:
As a Haskellite I vote for lazy. Logging isn't critical. If you don't have logging you care about performance and would appreciate us not loading the infrastructure if you don't use it.